llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgTransform/TransformOpTraits.h

```diff
@@ -0,0 +1,173 @@
+  static_assert(std::is_convertible<RetTy, LogicalResult>::value,
+                "expected transform function to return LogicalResult or "
+                "FailureOr<convertible-to-Operation*>");
+  for (Operation *target : targets) {
+    auto specificOp = dyn_cast<OpTy>(target);
+    if (!specificOp)
+      return failure();
+
+    auto result = transform(specificOp);
+    if (failed(appendTransformResultToVector(result, results)))
+      return failure();
+  }
+  return success();
+}
+} // namespace detail
+
+/// Trait implementing the TransformOpInterface for operations applying a
+/// transformation to a single operation handle and producing a single operation
+/// handle. The op must implement a method with one of the following signatures:
+///   - FailureOr<convertible-to-Operation*> applyToOne(OpTy)
+///   - LogicalResult applyToOne(OpTy)
+/// to perform a transformation that is applied in turn to all payload IR
+/// operations that correspond to the handle of the transform IR operation.
+/// In the functions above, OpTy is either Operation * or a concrete payload IR
+/// Op class that the transformation is applied to (NOT the class of the
+/// transform IR op).
+template <typename OpTy>
+class TargetableSingleOperandOpTrait
+    : public OpTrait::TraitBase<OpTy, TargetableSingleOperandOpTrait> {
+public:
+  /// Applies the transformation to each op from the only target and sets the
+  /// only result to correspond to the list of individual results.
+  LogicalResult apply(TransformResults &transformResults,
+                      TransformState &state) {
+    using TransformOpType = typename llvm::function_traits<decltype(
+        &OpTy::applyToOne)>::template arg_t<0>;
+    ArrayRef<Operation *> targets =
+        state.getPayloadOps(this->getOperation()->getOperand(0));
+    SmallVector<Operation *> results;
+    if (failed(detail::applyTransformToEach(
+            targets, results, [&](TransformOpType specificOp) {
+              return static_cast<OpTy *>(this)->applyToOne(specificOp);
+            })))
+      return failure();
+    if (OpTy::template hasTrait<OpTrait::OneResult>()) {
+      transformResults.set(
+          this->getOperation()->getResult(0).template cast<OpResult>(),
+          results);
+    }
+    return success();
+  }
+
+  /// Verifies that the op satisfies the requirements for this trait.
+  static LogicalResult verifyTrait(Operation *) {
+    static_assert(OpTy::template hasTrait<OpTrait::OneOperand>(),
+                  "expected single-operand op");
+    static_assert(OpTy::template hasTrait<OpTrait::OneResult>() ||
+                      OpTy::template hasTrait<OpTrait::ZeroResult>(),
+                  "expected zero- or single-result op");
+    return success();
+  }
+};
+
+template <typename OpTy>
+class FunctionalStyleTransformOpTrait
+    : public OpTrait::TraitBase<OpTy, FunctionalStyleTransformOpTrait> {
+public:
+  void getEffects(SmallVectorImpl<MemoryEffects::EffectInstance> &effects) {
```

Comment:
This is to allow running CSE and other things on the transform dialect itself, right?
Can you comment why you need this here?